File: javascript/03-composition/js/step-03.js
// Step 3: Weather Data Grid Visualization (with live Open-Meteo API)
// Visualizes 365 days of Syracuse, NY high temperatures as a grid of colored circles.
// Small circles = cold days, large circles = hot days. Colors use a weather-map style gradient.

// --- GLOBAL VARIABLES ---
let weatherData = null; // Will hold array of daily high temps (°F)
let loading = true;     // True while data is being fetched

// --- SETUP: p5.js entry point ---
function setup() {
  createCanvas(800, 800); // Create a square canvas
  background(255);        // White background
  textAlign(CENTER, CENTER); // Center text for loading/error
  textSize(32);
  fill(80);
  noLoop();               // Only draw when needed
  colorMode(RGB);         // Use RGB color mode
  ellipseMode(CENTER);    // Draw ellipses from the center
  // Start fetching weather data for Syracuse, NY
  fetchWeatherData();
}

// --- DRAW: p5.js main draw loop ---
function draw() {
  background(255);
  if (loading) {
    // Show loading message while data is being fetched
    fill(80);
    text('Loading Syracuse weather data...', width/2, height/2);
    return;
  }
  if (!weatherData) {
    // Show error if data failed to load
    fill(200,0,0);
    text('Failed to load weather data.', width/2, height/2);
    return;
  }

  // --- GRID LAYOUT ---
  // We want to fit all 365 days in a grid (20 columns x 19 rows)
  const cols = 20;
  const rows = ceil(weatherData.length / cols);
  const cellW = width / cols;
  const cellH = height / rows;

  // --- FIND MIN/MAX TEMPERATURES ---
  // We'll use these to map values to size and color
  const minTemp = min(weatherData);
  const maxTemp = max(weatherData);

  // --- DRAW EACH DAY AS A CIRCLE ---
  for (let i = 0; i < weatherData.length; i++) {
    let temp = weatherData[i]; // High temp for this day (°F)
    let col = i % cols;        // Column in the grid
    let row = floor(i / cols); // Row in the grid
    let x = col * cellW + cellW / 2; // Center x
    let y = row * cellH + cellH / 2; // Center y

    // --- MAP TEMPERATURE TO CIRCLE SIZE ---
    // Cold = small, hot = large
    let d = map(temp, minTemp, maxTemp, cellW * 0.3, cellW * 0.95);

    // --- MAP TEMPERATURE TO COLOR (weather map style) ---
    // We'll blend through blue → cyan → green → yellow → orange → red
    // using lerpColor for smooth transitions
    let cold = color(30, 90, 200);      // Blue (coldest)
    let cool = color(0, 200, 255);     // Cyan
    let mild = color(80, 220, 120);    // Green
    let warm = color(255, 230, 0);     // Yellow
    let hot = color(255, 120, 0);      // Orange
    let scorch = color(220, 30, 30);   // Red (hottest)
    // Normalize temp to 0-1 for blending
    let tNorm = map(temp, minTemp, maxTemp, 0, 1);
    let c;
    if (tNorm < 0.2) {
      c = lerpColor(cold, cool, tNorm / 0.2); // Blue to cyan
    } else if (tNorm < 0.4) {
      c = lerpColor(cool, mild, (tNorm - 0.2) / 0.2); // Cyan to green
    } else if (tNorm < 0.6) {
      c = lerpColor(mild, warm, (tNorm - 0.4) / 0.2); // Green to yellow
    } else if (tNorm < 0.8) {
      c = lerpColor(warm, hot, (tNorm - 0.6) / 0.2); // Yellow to orange
    } else {
      c = lerpColor(hot, scorch, (tNorm - 0.8) / 0.2); // Orange to red
    }

    // --- DRAW THE CIRCLE ---
    noStroke();
    fill(c);
    ellipse(x, y, d);
  }
}

// --- FETCH WEATHER DATA FROM OPEN-METEO API ---
// This function gets the last 365 days of daily high temps for Syracuse, NY
// and stores them in the global weatherData array.
function fetchWeatherData() {
  // Syracuse, NY: latitude=43.0481, longitude=-76.1474
  // Dates: 2024-05-04 to 2025-05-04 (change as needed)
  const url = 'https://archive-api.open-meteo.com/v1/archive?latitude=43.0481&longitude=-76.1474&start_date=2024-05-04&end_date=2025-05-04&daily=temperature_2m_max&temperature_unit=fahrenheit&timezone=America/New_York';
  // Use fetch() to get the data from the API
  fetch(url)
    .then(response => response.json()) // Parse the JSON response
    .then(data => {
      // Check if the data is valid and contains the temperature array
      if (data && data.daily && data.daily.temperature_2m_max) {
        weatherData = data.daily.temperature_2m_max; // Array of 365 numbers
        loading = false;
        redraw(); // Trigger draw() to update the canvas
      } else {
        weatherData = null;
        loading = false;
        redraw();
      }
    })
    .catch(err => {
      // If there's an error (e.g., network), show error message
      weatherData = null;
      loading = false;
      redraw();
    });
}

/*
  This sketch visualizes a year of Syracuse, NY high temperatures using live data from Open-Meteo.
  - Each day is a circle in a grid (20 columns x 19 rows).
  - Small blue circles = coldest days, large red/orange circles = hottest days.
  - Colors blend smoothly from blue (cold) to red (hot) like a weather map.
  - Try swapping in real data for your city by changing the latitude/longitude!
  - The code uses fetch() to get JSON data from the API, then draws the grid.
  - All mapping (size, color) is done with p5.js functions for clarity.
*/

File: python/03-composition/py/step-03.py
# Step 3: Weather Data Grid Visualization (Processing.py)
# ------------------------------------------------------
# Visualize a year of daily high temperatures as a grid of colored circles.
# For use in Processing IDE, Python mode.
#
# This sketch loads a CSV file (e.g. 'syracuse-temps.csv') with 365 daily high
# temperatures (Fahrenheit) and arranges them in a grid. Each day is shown as a
# circle whose size and color are mapped to the temperature: cold days are small
# and blue, hot days are large and red/orange. The color spectrum blends smoothly
# from blue (cold) to red (hot), like a weather map. A color legend is drawn at
# the bottom.
#
# TEACHER NOTES:
# - Use this sketch to discuss data visualization, color mapping, and abstraction.
# - Challenge students to use other weather data (rainfall, low temps, etc) or try a different city by changing the data file.
# - Ask: How does the color mapping affect your perception of the data? What other color palettes could you use?
# - Try changing the grid size, color stops, or mapping for new effects.
# - Encourage students to explain the code logic in their own words, especially how the data is used.
#
# CODE WALKTHROUGH:
# - grid_cols/grid_rows: sets the grid (20x19 fits 365 days).
# - color_stops: defines the color spectrum for mapping temperatures.
# - Loads the CSV file, handling both one-value-per-line and comma-separated formats.
# - For each day:
#     1. Map temperature to circle size and color.
#     2. Draw the circle in the correct grid position.
# - draw_legend() draws a color bar legend at the bottom.
# - If the data file is missing, an error message is shown.
#
# DATA FILE:
# - Place a CSV file (e.g. 'syracuse-temps.csv') in your sketch folder.
# - The file should contain 365 daily high temperatures (F), one per line or comma-separated.
#
# Example CSV (one value per line):
# 32.1
# 35.2
# 40.0
# ...
#
# Example CSV (single line, comma-separated):
# 32.1,35.2,40.0,...

import csv

grid_cols = 20
grid_rows = 19
filename = "syracuse-temps.csv"  # Place this file in your sketch folder

temps = []
min_temp = 999
max_temp = -999

# --- Color stops for weather map (blue to red) ---
color_stops = [
    color(0, 80, 220),    # Blue (coldest)
    color(0, 200, 255),   # Cyan
    color(80, 220, 120),  # Green
    color(240, 220, 100), # Yellow
    color(255, 140, 0),   # Orange
    color(220, 40, 40)    # Red (hottest)
]

def setup():
    """
    Processing.py entry point. Loads temperature data, computes min/max, and draws the grid.
    """
    global temps, min_temp, max_temp
    size(800, 800)
    background(255)
    noStroke()
    # --- Load temperature data (handles both one-value-per-line and comma-separated single line) ---
    try:
        with open(filename, 'r') as csvfile:
            reader = csv.reader(csvfile)
            for row in reader:
                # If the row is a single long string, split by comma
                if len(row) == 1 and ',' in row[0]:
                    values = row[0].split(',')
                    for v in values:
                        v = v.strip()
                        if v == '' or v.lower() == 'null':
                            continue
                        t = float(v)
                        temps.append(t)
                        if t < min_temp:
                            min_temp = t
                        if t > max_temp:
                            max_temp = t
                else:
                    for v in row:
                        v = v.strip()
                        if v == '' or v.lower() == 'null':
                            continue
                        t = float(v)
                        temps.append(t)
                        if t < min_temp:
                            min_temp = t
                        if t > max_temp:
                            max_temp = t
    except Exception as e:
        # Show error if data file is missing or invalid
        fill(200, 0, 0)
        textAlign(CENTER)
        textSize(24)
        text("Error loading data: " + str(e), width/2, height/2)
        return
    # --- Draw grid of circles ---
    cellW = width / grid_cols
    cellH = height / grid_rows
    for i in range(len(temps)):
        col = i % grid_cols
        row = i // grid_cols
        x = col * cellW + cellW / 2
        y = row * cellH + cellH / 2
        t = temps[i]
        # Map temp to size (d) and color (c)
        d = map(t, min_temp, max_temp, cellW * 0.3, cellW * 0.95)
        c = temp_to_color(t, min_temp, max_temp)
        fill(c)
        ellipse(x, y, d, d)
    # --- Add color legend ---
    draw_legend(min_temp, max_temp)

def temp_to_color(t, tmin, tmax):
    """
    Map a temperature value t to a color using the color_stops list.
    The mapping is smooth (interpolated) between color stops.
    """
    norm = (t - tmin) / float(tmax - tmin)
    n = len(color_stops) - 1
    idx = int(norm * n)
    idx = min(idx, n - 1)
    frac = (norm * n) - idx
    c1 = color_stops[idx]
    c2 = color_stops[idx + 1]
    return lerpColor(c1, c2, frac)

def draw_legend(tmin, tmax):
    """
    Draw a horizontal color bar legend at the bottom of the canvas, showing the temperature-color mapping.
    """
    legendW = 400
    legendH = 24
    x0 = width/2 - legendW/2
    y0 = height - 40
    for i in range(legendW):
        norm = i / float(legendW - 1)
        t = tmin + norm * (tmax - tmin)
        c = temp_to_color(t, tmin, tmax)
        stroke(c)
        line(x0 + i, y0, x0 + i, y0 + legendH)
    noStroke()
    fill(50)
    textAlign(CENTER)
    textSize(16)
    text(str(int(tmin)) + u"°F", x0, y0 + legendH + 18)
    text(str(int(tmax)) + u"°F", x0 + legendW, y0 + legendH + 18)
    text("Coldest", x0, y0 + legendH + 34)
    text("Hottest", x0 + legendW, y0 + legendH + 34)